pkg/controller/machineset.go

```diff
@@ -664,6 +666,8 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 		numOfInactiveMachines = len(inactiveMachines)
 		errCh                 = make(chan error, numOfInactiveMachines)
 	)
+	defer close(errCh)
```

Comment:
Just from readability perspective, I would refactor it as:
```suggestion
	wg.Add(numOfInactiveMachines)
	for _, machine := range inactiveMachines {
		go c.prepareMachineForDeletion(machine, machineSet, &wg, &errCh)
	}
	wg.Wait()
	close(errCh)
	err, ok := <-errCh
	if !ok {
	 return nil
	}
	return err
	
```

Reply:
> ```go
> close(errCh)
> err, ok := <-errCh
> ```

Won't work :-)

Reply:
It will work, https://play.golang.org/p/Oz3XDGZpUM8
Readability is perspective based. Your choice.

Reply:
👍 You are right. My bad.

My comments were not based on readability. I do prefer the error check standard pattern over the switch check.

Reply:
BTW regarding readability, I think we don't really need the call to `prepareMachineForDeletion` as goroutines in the first place. I think we are being over-aggressive with asynchronous calls. But that is for another day.

Reply:
Closing a channel before reading from it won't work:
https://play.golang.org/p/F1Z7z0Zs5a3
In @swapnilgm's go playground it works because the channel is read before the go routine could close it. It is also a potential race.

Reply:
Even in @swapnilgm's suggested changes, there could be a race. The channel could be closed before any of the go routines complete/scheduled. I prefer the `defer` way of closing the channel.

Reply:
Just saw: `wg.Wait()` would take care of the potential race. However, closing the channel before reading would not work.

Reply:
Closing of channel is developers reponsibility. Developer has to take care of race condition and all. Here, as you already noticed `wg.Wait()` takes care of it. 

> However, closing the channel before reading would not work.

The closing of channel before reading is totally fine by golng design. Its been used many places in golang based project. Nothing wrong in it. Only writing to channel is not allowed after closing the channel.

Reply:
what about this: https://play.golang.org/p/F1Z7z0Zs5a3
Doesn't seem to work here. Am I missing something?

Reply:
Ofcourse its race condition. Its your responsibility to handle it, but you missed it. If the go routine you had created won the race, then it would been paniced.
check : https://play.golang.org/p/2J7FOtazkXd . I simply added `sleep` so that go routing gets executed first post closing of channel.

Reply:
Writing  to unbuffered channel  is blocking call. Try making it buffered channel : https://play.golang.org/p/2J7FOtazkXd

Reply:
I am keeping the defer logic as it is. However, I am changing the receiving channel to write only as pointed out by Swapnil. Refer - https://github.com/gardener/machine-controller-manager/pull/341/commits/9c50392cdb691bfbafc51a6dcb8496256f64e4a6.